### PR TITLE
[data registry] form linking multiple case types

### PR DESCRIPTION
**Pull request by snopoke**
## Product Description
Bug fix for using form linking in conjunction with data registries and multiple case types for case search.

## Technical Summary
With multiple case types in play the datum matching for data registry queries was failing. 

There are two commits with fixes related to this which address the issue in different parts of the code. Other commits are non-functional changes / refactors.

### [infer case type for 'next_meta'](https://github.com/dimagi/commcare-hq/commit/cb6bbd8e7b84bb0db122d546fca4480139c73fc1)

Normally the case types of workflow datums are parsed from the datum nodeset. In cases where this isn't possible the case type is set to that of the module (this happens in a post construction step). This step was being skipped for datums used when constructing `WorkflowQueryMeta` objects. The impact of this was that the datum matching was failing due to the missing case type in the datum passed to `WorkflowQueryMeta`.

Jira: https://dimagi-dev.atlassian.net/browse/USH-1446

### [support multiple data elements with the same ID](https://github.com/dimagi/commcare-hq/commit/0e050caa118bb75484b829718a7ff2bdddede952)

This commit fixes a bug where data elements in query datums were overwriting previously seen elements with the same ID. This was impacting both 'multiple case types' as well as 'additional registry cases'.

Jira: https://dimagi-dev.atlassian.net/browse/USH-1450

**Best reviewed by commit**

## Feature Flag
DATA_REGISTRY, USH_CASE_CLAIM_UPDATES

### Safety story
The code is well covered by tests and the changes only impact feature flagged areas.

### Automated test coverage
Added new tests to cover the changes.

### QA Plan
This feature is not currently in use by any live applications. The USH delivery team is in the process of testing / building these features into their apps.

### Migrations
NA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/app_manager/tests/util.py [hunk 1]
```diff
@@ -25,14 +25,7 @@ class TestXmlMixin(TestFileMixin):
...
     def assertXmlPartialEqual(self, expected, actual, xpath):
-        """
-        Extracts a section of XML using the xpath and compares it to the expected
-
-        Extracted XML is placed inside a <partial/> element prior to comparison.
-        """
-        expected = parse_normalize(expected)
-        actual = extract_xml_partial(actual, xpath)
-        self.assertXmlEqual(expected, actual, normalize=False)
+        assert_xml_partial_equal(expected, actual, xpath)
```
Comment (esoergel) [inline corehq/apps/app_manager/tests/util.py:28]: :+1:

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py [hunk 2]
```diff
@@ -200,7 +200,7 @@ def _remote_request_query_datums(self):
         additional_types = set(self.module.search_config.additional_case_types) - {self.module.case_type}
         datums = [
             QueryData(key='case_type', ref=f"'{case_type}'")
-            for case_type in [self.module.case_type] + list(additional_types)
+            for case_type in [self.module.case_type] + sorted(additional_types)
```
Comment (esoergel) [inline corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:203]: Is this a concept that would make sense to abstract to ensure it's consistent?  (feel free to say no)

**Review by esoergel — APPROVED** [review]

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:203]: I did consider that but there are subtle difference in how the data is passed in that would make it less useful. Though perhaps just having a function would make it clear that they should be consistent. I'll do a follow up PR.

Comment (snopoke) [reply corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:203]: https://github.com/dimagi/commcare-hq/pull/30829

**Pull request merged**